keps/sig-node/2570-memory-qos/README.md

```diff
@@ -0,0 +1,164 @@
+# KEP-2570: Support Memory QoS using cgroups v2
+## Summary
+Support memory qos in pod qos class using cgroups v2.
+
+## Motivation
+In traditional cgroups v1 implement in Kubernetes, pod qos class and qos manager can only limit cpu resources, such as cpu_shares, memory qos is not yet implemented. cgroups v2 brings new capabilities for memory controller and it would help Kubernetes enhance memory isolation quality.
+
+### Goals
+- Support memory qos for pod qos classes
+
+### Non-Goals
+- additional qos design
+- support other resources qos
+
+## Proposal
+Use memory controller interfaces of cgroups v2 to support memory qos of pod qos class.
+
+Kubernetes cgroup manager uses `memory.limit_in_bytes` in cgroups v1 to limit memory capacity and uses `oom_scores` to determine order of killing container process when OOM occurs. 
+
+Above implement has following disadvantages:
+- `Guaranteed` memory can not be fully reserved, page cache is at risk of being recycled
+- `Burstable` memory allocation overcommit ( request < limit ) could increase OOM possibility when kernel memory runs out
+
+cgroups v2 introduces a better way to limit and guarantee memory quality.
+
+| File | Description |
+| -------- | -------- |
+| memory.high    | memory.high is the memory usage throttle limit. This is the main mechanism to control a cgroup’s memory use. If a cgroup's memory use goes over the high boundary specified here, the cgroup’s processes are throttled and put under heavy reclaim pressure. The default is max, meaning there is no limit. |
+| memory.max | memory.max is the memory usage hard limit, acting as the final protection mechanism: If a cgroup's memory usage reaches this limit and can't be reduced, the system OOM killer is invoked on the cgroup. Under certain circumstances, usage may go over the memory.high limit temporarily. When the high limit is used and monitored properly, memory.max serves mainly to provide the final safety net. The default is max. |
+| memory.low | memory.low is the best-effort memory protection, a “soft guarantee” that if the cgroup and all its descendants are below this threshold, the cgroup's memory won't be reclaimed unless memory can’t be reclaimed from any unprotected cgroups. |
+| memory.min | memory.min specifies a minimum amount of memory the cgroup must always retain, i.e., memory that can never be reclaimed by the system. If the cgroup’s memory usage reaches this low limit and can’t be increased, the system OOM killer will be invoked. |
+
+This proposal uses `memory.min` to retain request / limit memory for `Guaranteed` Pod. And for `Burstable` Pod, it uses `memory.low` to achieve best-effort memory protection as well as using `memory.high` to throttle allocation of memory overcommit.
```

Comment:
One thing that's still not clear to me, is the proposal here suggesting that the memory limit will map to `memory.high`? If so, does it mean that the `memory.max` will be unset? 

According to the [kernel cgroup v2 docs](https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html#memory-interface-files) for `memory.high`

> Going over the high limit never invokes the OOM killer 

As a result, some questions I have:

1) what is the behavior if only memory.high will be set but `memory.max` will be unset?
2) The OOM on crossing memory.limit is current behavior in cgroup v1 and changing that behavior may be surprising to end users. Is this change intentional here? 
  2a) If so, I'm a bit concerned that since OOM killer won't be invoked it will lead to heavy reclaim and instead result in page cache/disk thrashing. Is that case / what are your thoughts here?

Reply:
> One thing that's still not clear to me, is the proposal here suggesting that the memory limit will map to `memory.high`? If so, does it mean that the `memory.max` will be unset?
> 
> According to the [kernel cgroup v2 docs](https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html#memory-interface-files) for `memory.high`
> 
> > Going over the high limit never invokes the OOM killer
> 
> As a result, some questions I have:
> 
> 1. what is the behavior if only memory.high will be set but `memory.max` will be unset?
It still needs to set `memory.max = limits.memory` to make cgroup oom killer work.

> 2. The OOM on crossing memory.limit is current behavior in cgroup v1 and changing that behavior may be surprising to end users. Is this change intentional here?
>    2a) If so, I'm a bit concerned that since OOM killer won't be invoked it will lead to heavy reclaim and instead result in page cache/disk thrashing. Is that case / what are your thoughts here?
Yes, you are right, we need keep current behavior of limits.memory. Maybe we should set `memory.high = requests.memory` in burstable pods, it also makes sense that memory use goes over requests would be throttled and put under reclaim pressure.